src/pyicloud_ipd/base.py

```diff
@@ -192,93 +192,15 @@ def authenticate(self, force_refresh:bool=False, service:Optional[Any]=None) ->
-            uname, A = usr.start_authentication()
-            data = {
-                'a': base64.b64encode(A).decode(),
-                'accountName': uname,
-                'protocols': ['s2k', 's2k_fo']
-            }
-
-            try:
-                response = self.session.post("%s/signin/init" % self.AUTH_ENDPOINT, data=json.dumps(data), headers=headers)
-                if response.status_code == 401:
-                    raise PyiCloudAPIResponseException(response.text, str(response.status_code))
-            except PyiCloudAPIResponseException as error:
-                msg = "Failed to initiate srp authentication."
-                raise PyiCloudFailedLoginException(msg, error) from error
-
-            # Step 2: server sends public key B, salt, and c to client
-            body = response.json()
-            salt = base64.b64decode(body['salt'])
-            b = base64.b64decode(body['b'])
-            c = body['c']
-            iterations = body['iteration']
-            protocol = body['protocol']
-
-            # Step 3: client generates session key M1 and M2 with salt and b, sends to server
-            srp_password.set_encrypt_info(protocol, salt, iterations)
-            m1 = usr.process_challenge( salt, b )
-            m2 = usr.H_AMK
-
-            data = {
-                "accountName": uname,
-                "c": c,
-                "m1": base64.b64encode(m1).decode(),
-                "m2": base64.b64encode(m2).decode(),
-                "rememberMe": True,
-                "trustTokens": [],
-            }
-
-            if self.session_data.get("trust_token"):
-                data["trustTokens"] = [self.session_data.get("trust_token")]
-
             try:
-                response = self.session.post(
-                    "%s/signin/complete" % self.AUTH_ENDPOINT,
-                    params={"isRememberMeEnabled": "true"},
-                    data=json.dumps(data),
-                    headers=headers,
-                )
-                if response.status_code == 409:
-                    # requires 2FA
-                    pass
-                elif response.status_code == 412:
-                    # non 2FA account returns 412 "precondition no met"
-                    headers = self._get_auth_headers()
-                    response = self.session.post(
-                        "%s/repair/complete" % self.AUTH_ENDPOINT,
-                        data=json.dumps({}),
-                        headers=headers,
-                    )
-                elif response.status_code >= 400 and response.status_code < 600:
-                    raise PyiCloudAPIResponseException(response.text, str(response.status_code))
-            except PyiCloudAPIResponseException as error:
-                msg = "Invalid email/password combination."
-                raise PyiCloudFailedLoginException(msg, error) from error
+                self._authenticate_srp()
+            except PyiCloudFailedLoginException as error:
```

Comment:
I wonder if there is anything in payload we can use as a flag for srp rather than try-catch.

If try-catch is the only way, what other exceptions may be incorrectly channelled to non-srp path?

Reply:
For the account I tried, the response for srp signin/complete is:
```
('Invalid email/password combination.', PyiCloudAPIResponseException('{\n  "serviceErrors" : [ {\n    "code" : "-20283",\n    "message" : "Enter the email or phone number and password for your Apple\xa0Account.",\n    "suppressDismissal" : false\n  } ]\n} (401)'))
```

Which is the same as when the password is incorrect.

This also means if the password is incorrect, it will be incorrectly channelled to non-srp path.